### ISSUE TITLE
Cannot open seccomp filter: permission denied when installing snap

### ISSUE THREAD

**Issue by 3nt3** [certbot/certbot#9372]
## My operating system is (include version):
Linux 5.4.0-122-generic - Ubuntu 20.04 LTS

## I installed Certbot with (snap, OS package manager, pip, certbot-auto, etc):
snap

## I ran this command and it produced this output:
```
$ snap install --classic certbot
error: cannot perform the following tasks:
- Run configure hook of "certbot" snap if present (run hook "configure": cannot open seccomp filter /var/lib/snapd/seccomp/bpf//snap.certbot.hook.configure.bin: Permission denied)
```

## Certbot's behavior differed from what I expected because:
I can't install it

**Comment by bmw** [comment]
Figuring out exactly where the problem is here might be tricky... Does your system noteworthy in any way? For instance, what CPU architecture does it use (i386, AMD64, some version of ARM, etc.)?

Also just to check, did you follow all of the instructions at https://certbot.eff.org/instructions?ws=webproduct&os=ubuntufocal? If not, I wonder if steps like upgrading snapd would solve the problem for you.

**Comment by 3nt3** [comment]
I upgraded snap and also my whole system, it runs on x86_64 nothing special about it I htink.

**Comment by 3nt3** [comment]
Also I reinstalled snapd completely but it did nothing

**Comment by bmw** [comment]
I honestly have no idea what's going on here unfortunately. When I [search google for the generic part of this error message](https://www.google.com/search?q=%22cannot+open+seccomp+filter%22), at least as of writing this the only results are your posts and one person having trouble with another snap on a Raspberry Pi. The reason I asked about the peculiarities of your system is due to posts like that and [people having had trouble with snaps in virtualized environments in the past](https://github.com/certbot/website/issues/636).

I think it's hard to debug where exactly in the system the problem is here. If anyone is able to dig into it and provide more information, that'd be great, but for now I recommend installing Certbot through Docker or pip as described [here](https://eff-certbot.readthedocs.io/en/stable/install.html#alternate-installation-methods).

**Comment by gbnewby** [comment]
I found this issue report because I had the same problem. I saw this posting that mentioned permissive file permissions might be the issue: https://community.letsencrypt.org/t/running-certbot-failed-cannot-find-installed-snap/143105

My situation is that /var was actually a symlink to another filesystem. I needed to do this after the root (/) filesystem filled up. Symlinks are always mode 777, even though what they point to might be 755 or anything else.

I cleaned up some space and moved /var back to / so it was not a symlink any more. The error above then went away and I was able to install certbot and hello-world without any errors or warnings. "snap list" no longer showed "broken" installations, and overall things worked as expected.

I don't know whether this was the original poster's problem or not, but I had the exact same error message, and moving /var from a symlink back to a regular directory solved the problem.

Based on this, perhaps the snap code could be updated to better handle symlinks. While symlinks are not ideal for /var, it seems that snap is not actually checking what it  intends, which is the perms on the link target.

Thanks to the snap team for delivering a great product!

**Comment by 3nt3** [comment]
Yes, `/var/` is symlinked to another drive! I really don't know how you figured this out but I can't really fix it either since I really don't have enough disk space

**Comment by bmw** [comment]
@gbnewby, thanks so much for figuring that out and sharing what you found.

I created https://forum.snapcraft.io/t/improve-handling-of-symbolic-links/31726 to talk with the snap team about improving the experience here. Since I think this is a snap problem rather than a Certbot problem, I'm going to close this issue, but please let me know if you think I should reconsider.

@3nt3, while doing my own research, I read a number of times that replacing the symlink with a bind mount is likely to solve your problem.

**Comment by gbnewby** [comment]
Thanks for this, Brad. Based on the response from the apparmor team, it
seems there is a Certbot "documentation solution" to the issue.

This is because it's not really an issue with snapd, but with how apparmor
implements symlink protection.

This page says that the issue won't be fixed:
https://bugs.launchpad.net/apparmor/+bug/1485055

However, the page offers a clear solution that should work for people who
encountered the issue when trying to get Certbot running.

Therefore, my suggestion is to consider adding a mention of the problem on
the Certbot documentation pages, and the solution (or a link to the
Launchpad page). Because Certbot does a great job at being very easy to
install across various platforms, I don't think it's inconsistent with the
support philosophy to mention that this particular problem might occur, and
that while it originates two layers down from Certbot, there is a
straightforward solution.

Thanks for your consideration, and for keeping after this issue. It seems
like it's not that common, but I believe it requires a fair amount of
syadmin/devops sophistication to diagnose & mitigate. I would hate to think
that there are potential Certbot users who were blocked from using Certbot
due to this issue.

**Comment by bmw** [comment]
Fair. I think the further up the dependency tree this problem is fixed, the more people will benefit, but it's seeming less and less likely this problem is going to get fixed upstream.

I'm open to ideas on how to solve this problem with documentation on our end, but I think the problem is rare enough that:

1. This probably isn't something the Certbot team at EFF is going to prioritize anytime soon.
2. I don't think it's worth putting it in our main instructions at https://certbot.eff.org/instructions. Maybe it can go in https://eff-certbot.readthedocs.io/en/stable/install.html or maybe we could start a troubleshooting page under https://eff-certbot.readthedocs.io/? (Our readthedocs site is linked to through https://certbot.eff.org/docs on pages like https://certbot.eff.org/pages/help.)

**Comment by gbnewby** [comment]
Hi Brad. Thanks for this. I spent some time looking through the available
documentation you referenced.

I think that the best place to document this issue seems to be this page:
https://eff-certbot.readthedocs.io/en/stable/

My suggestion would be to add a section towards the end, "Installation
issues." I bet there are more issues that people have encountered than the
one reported in this github issue we are discussing, and I didn't find
anyplace else (other than this github page) where issues are addressed.
Having a section that addresses issues seems very reasonable to me, on the
installation instructions page.

At first I thought maybe this was more of an FAQ item, but the Certbot FAQ
is mostly about general functionality and purpose. It's not technical
(other than one item about use of --manual to install wildcard
certificates, which might also be better on the Installation page above
under "issues.").

Potential text:

Issue: Snap fails with "permission denied" error

When running this command, a "permission denied" error may result:

$ snap install --classic certbot

This may be due to the /var directory being a symbolic link to a separate
file system than the / (root) partition. The problem is not caused by
Certbot, but by a reliance of snap on the apparmor security utility. The
solution is to either have /var be a mounted partition (rather than a
symbolic link), or to add the linked partition to
/etc/apparmor.d/tunables/alias (or whatever is the correct location for
your system). Additional information about the issue may be found here:
https://bugs.launchpad.net/apparmor/+bug/1485055



I hope this is helpful. I won't be upset if EFF decides not to add this to
the documentation (since, after all, I eventually managed to find a fix).
Anyone who puts /var on a different partition probably has at least a basic
understanding of how to run systems and diagnose problems. Though of
course, the person installing Certbot (say, a webmaster or developer) might
not be the sysadmin who set up the system.

Thanks for your consideration. Greg
